include/crow/middlewares/session.h

```diff
@@ -0,0 +1,633 @@
+            }
+
+            // check this is a valid entry before loading
+            if (!store_.contains(session_id)) return;
+
+            auto node = std::make_shared<session::CachedSession>();
+            node->session_id = session_id;
+            node->referrers = 1;
+
+            try
+            {
+                store_.load(*node);
+            }
+            catch (...)
+            {
+                CROW_LOG_ERROR << "Exception occurred during session load";
+                return;
+            }
+
+            ctx.node = node;
+            cache_[session_id] = node;
+        }
+
+        template<typename AllContext>
+        void after_handle(request& /*req*/, response& /*rsp*/, context& ctx, AllContext& all_ctx)
+        {
+            lock l(*mutex_);
+            if (!ctx.node || --ctx.node->referrers > 0) return;
+            ctx.node->requested_refresh |= ctx.node->session_id == "";
+
+            // generate new id
+            if (ctx.node->session_id == "")
+            {
+                // check for requested id
+                ctx.node->session_id = std::move(ctx.node->requested_session_id);
+                if (ctx.node->session_id == "")
+                {
+                    ctx.node->session_id = utility::random_alphanum(id_length_);
+                }
+            }
+            else
+            {
+                cache_.erase(ctx.node->session_id);
+            }
+
+            if (ctx.node->requested_refresh)
+            {
+                auto& cookies = all_ctx.template get<CookieParser>();
+                store_id(cookies, ctx.node->session_id);
+            }
+
+            try
+            {
+                store_.save(*ctx.node);
+            }
+            catch (...)
+            {
+                CROW_LOG_ERROR << "Exception occurred during session save";
+                return;
+            }
+        }
+
+    private:
+        std::string next_id()
+        {
+            std::string id;
+            do
+            {
+                id = utility::random_alphanum(id_length_);
+            } while (store_.contains(id));
+            return id;
+        }
+
+        std::string load_id(const CookieParser::context& cookies)
+        {
+            auto session_raw = cookies.get_cookie(cookie_.name());
+
+            auto pos = session_raw.find(":");
+            if (pos <= 0 || pos >= session_raw.size() - 1) return "";
+
+            auto session_id = session_raw.substr(0, pos);
+            auto session_hmac = session_raw.substr(pos + 1, session_raw.size() - pos - 1);
+
+            return hash_id(session_id) == session_hmac ? session_id : "";
+        }
+
+        void store_id(CookieParser::context& cookies, const std::string& session_id)
+        {
+            cookie_.value(session_id + ":" + hash_id(session_id));
+            cookies.set_cookie(cookie_);
+        }
+
+        std::string hash_id(const std::string& key) const
+        {
+            sha1::SHA1 s;
```

Comment:
It is to some extend... Crow uses SHA1 for websocket handshakes. 
We can remove the possibility to set custom keys altogether. This will remove the requirement for signing. I actually think now this is a more viable option.